### PR TITLE
add metrics

### LINKED ISSUE #3219: expose  metrics
**What's it about?**

expose metric

**What's the reason why we need it?**

I believe this is an important feature for Kubesphere. There're a few use cases:

* monitor http latency, http code , go gc duration and something else

Please leave your comments below if there's anyone agrees with me. Or just give me a thumb up.

**Area Suggestion**
/area apiserver
/kind feature-request

### PR DESCRIPTION
**Pull request by xyz-li**
Signed-off-by: LiHui <andrewli@yunify.com>

**What type of PR is this?**
/kind feature
/area apiserver

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #3219

**How to add new metric**:

1. register metric
 ```
package xxxx

import (
	compbasemetrics "k8s.io/component-base/metrics"
	"kubesphere.io/kubesphere/pkg/utils/metrics"
)

var (
	appTemplateCreationCounter = compbasemetrics.NewCounterVec(
		&compbasemetrics.CounterOpts{
			Name:           "application_template_creation",
			Help:           "Counter of application template creation broken out for each workspace, name and create state",
			StabilityLevel: compbasemetrics.ALPHA,
		},
		[]string{"workspace", "name", "state"},
	)
)

func init() {
	Register()
}

func Register() {
	metrics.MustRegister(appTemplateCreationCounter)
}

```

2. export metric

```
package xxxx
func (c *appTemplateOperator) CreateApp(request *CreateAppRequest) (*CreateAppResponse, error) {
	appTemplateCreationCounter.WithLabelValues("workspace", "Name", "failed").Inc()
	
        return &CreateAppResponse{
		AppID:     resp.GetAppId().GetValue(),
		VersionID: resp.GetVersionId().GetValue(),
	}, nil
}
```

### CODE REVIEW

File: pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -167,6 +173,24 @@ func (s *APIServer) PrepareRun(stopCh <-chan struct{}) error {
...
+	reqInfo, exists := request.RequestInfoFrom(r.Request.Context())
+	if exists && reqInfo.APIGroup != "" {
+		RequestCounter.WithLabelValues(reqInfo.Verb, reqInfo.APIGroup, reqInfo.APIVersion, reqInfo.Resource, strconv.Itoa(response.StatusCode())).Inc()
+		elapsedSeconds := time.Now().Sub(start).Seconds()
+		RequestLatencies.WithLabelValues(reqInfo.Verb, reqInfo.APIGroup, reqInfo.APIVersion, reqInfo.Resource).Observe(elapsedSeconds)
+	}
+}
+
+func (s *APIServer) installAPI() {
+	if s.Config.MetricsOptions != nil && s.Config.MetricsOptions.Enable {
```
Comment (zryfish) [inline pkg/apiserver/apiserver.go:188]: installAPI is not specific, and why not put inside `installKubeSphereAPIs`

File: pkg/apiserver/config/config.go [hunk 2]
```diff
@@ -96,6 +97,7 @@ type Config struct {
 	AuditingOptions       *auditingclient.Options                    `json:"auditing,omitempty" yaml:"auditing,omitempty" mapstructure:"auditing"`
 	AlertingOptions       *alerting.Options                          `json:"alerting,omitempty" yaml:"alerting,omitempty" mapstructure:"alerting"`
 	NotificationOptions   *notification.Options                      `json:"notification,omitempty" yaml:"notification,omitempty" mapstructure:"notification"`
+	MetricsOptions        *metrics.Options                           `json:"metrics,omitempty" yaml:"metrics,omitempty" mapstructure:"metrics"`
```
Comment (zryfish) [inline pkg/apiserver/config/config.go:100]: I think we can just enable metrics by default

**Comment by zryfish** [comment]
Can you write a guide for our developers, how to add metrics on their own? For example, how to add metrics for devops module?

**Comment by xyz-li** [comment]
**How to add new metric**:

1. Define your metric in the  file `metric.go` of you package or somewhere else. 
metric.go
 ```
 package application

import (
	compbasemetrics "k8s.io/component-base/metrics"
	"kubesphere.io/kubesphere/pkg/utils/metrics"
 )

var (
      // add a counter metric
	appTemplateCreationCounter = compbasemetrics.NewCounterVec(
		&compbasemetrics.CounterOpts{
			Name:           "application_template_creation",  // metric name
			Help:           "Counter of application template creation broken out for each workspace, name and create state",   // metric help info
			StabilityLevel: compbasemetrics.ALPHA,  //represents the API guarantees for a given defined metric
		},
		[]string{"workspace", "name", "state"},  // the  labels of the metric
	)
     
     // add a histogram metric
	requestLatencies = compbasemetrics.NewHistogramVec(
		&compbasemetrics.HistogramOpts{
			Name: "ks_server_request_duration_seconds",
			Help: "Response latency distribution in seconds for each verb, group, version, resource",
			Buckets: []float64{0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9},
			StabilityLevel: compbasemetrics.ALPHA,
		},
		[]string{"verb", "group", "version", "resource"},  // the  labels of the metric
	)
 )

```

2. Register the metric in the file `metric.go`

```
package application
 import (
  "kubesphere.io/kubesphere/pkg/utils/metrics"
 )

func init(){
  Register()
}

func Register() {
	metrics.MustRegister(appTemplateCreationCounter)
       metrics.MustRegister(requestLatencies)
}
```


3. Export the metric

```
package application
func (c *appTemplateOperator) CreateApp(request *CreateAppRequest) (*CreateAppResponse, error) {
      // export app template create metric with label values
	appTemplateCreationCounter.WithLabelValues("workspace", "Name", "failed").Inc()
     //export request latency
       requestLatencies.WithLabelValues( "POST", "groupName", "versionName", "resourceName").Observe(elapsedSeconds)
        return  
}
```

**Comment by zryfish** [comment]
@xyz-li conflicts now. Could you please reorganize the above document, and submit it to the community repo?

File: pkg/apiserver/metric.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		[]string{"verb", "group", "version", "resource"},
+	)
+
+	metricsList = []compbasemetrics.Registerable{
+		RequestCounter,
+		RequestLatencies,
+	}
+)
+
+func register() {
```
Comment (wansir) [inline pkg/apiserver/metric.go:38]: Ambiguous method name.

**Comment by zryfish** [comment]
> @xyz-li conflicts now. Could you please reorganize the above document, and submit it to the community repo?

Any update for this?

**Pull request merged**